### PR TITLE
Hot code replace implementation

### PR DESCRIPTION
**Pull request by yaohaizh**
@andxu @testforstephen @akaroml

### CODE REVIEW

File: com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/DebugUtility.java [hunk 1]
```diff
@@ -313,6 +315,26 @@ public static void resumeThread(ThreadReference thread) {
...
+     *
+     * @param thread
+     *            the thread reference
+     * @return the list of the computed stack frame model
+     */
+    public static List<IStackFrame> computeStackFrame(ThreadReference thread) {
+        List<IStackFrame> result = new ArrayList<IStackFrame>();
+        try {
+            List<StackFrame> rawFrames = thread.frames();
+            for (StackFrame frame : rawFrames) {
```
Comment (andxu) [inline com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/DebugUtility.java]: list all  stack frame is very time-consuming, currently stack frame is listed using the pagination way.

File: com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/JavaHotCodeReplaceProvider.java [hunk 2]
```diff
@@ -0,0 +1,823 @@
...
+
+    private void filterNotLoadedTypes(List<IResource> resources, List<String> qualifiedNames) {
+        for (int i = 0, numElements = qualifiedNames.size(); i < numElements; i++) {
+            String name = qualifiedNames.get(i);
+            List<ReferenceType> list = getjdiClassesByName(name);
+            if (list.isEmpty()) {
+                // If no classes with the given name are loaded in the VM, don't
+                // waste
+                // cycles trying to replace.
+                qualifiedNames.remove(i);
```
Comment (andxu) [inline com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/JavaHotCodeReplaceProvider.java]: cannot remove element during iteration.

File: com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/handler/SetBreakpointsRequestHandler.java [hunk 3]
```diff
@@ -35,16 +41,31 @@
...
+    private boolean isInitialized = false;
+
     @Override
     public List<Command> getTargetCommands() {
         return Arrays.asList(Command.SETBREAKPOINTS);
     }
 
     @Override
     public CompletableFuture<Response> handle(Command command, Arguments arguments, Response response, IDebugAdapterContext context) {
+        synchronized (this) {
```
Comment (testforstephen) [inline com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/handler/SetBreakpointsRequestHandler.java]: Change `getTargetCommands()` to
<pre>
    public List<Command> getTargetCommands() {
           return Arrays.asList(Command.SETBREAKPOINTS, Command.CONFIGURATIONDONE);
    }
</pre>
Then put HCR initialization work in configurationDone handler.
<pre>
   if (command == Command.CONFIGURATIONDONE) {
           ... initialize HCR Provider.
   }
</pre>

File: com.microsoft.java.debug.plugin/META-INF/MANIFEST.MF [hunk 4]
```diff
@@ -13,7 +13,8 @@ Require-Bundle: org.eclipse.core.runtime,
  org.eclipse.debug.core,
  org.eclipse.jdt.debug,
  org.eclipse.jdt.core,
- org.eclipse.jdt.ls.core
+ org.eclipse.jdt.ls.core,
+ org.eclipse.core.filesystem;bundle-version="1.7.0"
```
Comment (testforstephen) [inline com.microsoft.java.debug.plugin/META-INF/MANIFEST.MF]: Better not to add the specific bundle version.

**Pull request closed without merging**